File: ruby/bob/bob.rb
module Bob
  def self.hey(remark)
    if yell?(remark)
      'Whoa, chill out!'
    elsif question?(remark)
      'Sure.'
    elsif nothing?(remark)
      'Fine. Be that way!'
    else
      'Whatever.'
    end
  end

  def self.yell?(remark)
    /[A-Z]+/.match(remark) && !/[a-z]+/.match(remark)
  end

  def self.question?(remark)
    remark.strip.end_with?('?')
  end

  def self.nothing?(remark)
    remark.strip.empty?
  end
end

module BookKeeping
  VERSION = 1
end

File: elixir/bob/bob.exs
defmodule Bob do
  def hey(input) do
    cond do
      silent?(input) -> "Fine. Be that way!"
      yell?(input) and question?(input) -> "Calm down, I know what I'm doing!"
      question?(input) -> "Sure."
      yell?(input) -> "Whoa, chill out!"
      true -> "Whatever."
    end
  end

  def silent?(input) do
    String.trim(input) == ""
  end

  def question?(input) do
    String.ends_with?(input, "?")
  end

  def yell?(input) do
    letters?(input) and String.upcase(input) == input
  end

  def letters?(input) do
    String.match?(input, ~r/\p{L}/)
  end
end
